internal/state/dataplane/configuration.go

```diff
@@ -333,6 +327,23 @@ func (hpr *hostPathRules) upsertListener(l *graph.Listener) {
 	}
 }
 
+func getHostnames(r *graph.Route, l *graph.Listener) []string {
+	var hostnames []string
+	for _, h := range r.Source.Spec.Hostnames {
+		if _, exist := l.AcceptedHostnames[string(h)]; exist {
+			hostnames = append(hostnames, string(h))
+		}
+	}
+
+	if len(r.Source.Spec.Hostnames) == 0 {
```

Comment:
Can you add a comment here that explains this case? It took me a bit to figure it out

Reply:
Is it possible to eliminate this case altogether? It seems weird that we have to check for this case twice -- once in the `graph` package and once in the `dataplane` package. 

What do you think about adding a `Hostnames` field to `graph.Route`? If an empty hostname is provided, the field will contain the wildcard hostname or the listener hostname; otherwise, it will be set to `r.Source.Spec.Hostnames`. Then in this function, we would only need one for loop that ranges over `r.Hostnames`. Would that work?

Eventually, we'd like to remove the `Source` field from the graph objects, so this would be one step in that direction.

Reply:
Yeah let me try that approach.

Reply:
Do you think that would be confusing at all if the HTTPRoute has no hostnames, but the Route object has a list of hostnames (either the listener hostname or wildcard)? Or are we okay with that since it's just for internal plumbing?

Reply:
> Is it possible to eliminate this case altogether? It seems weird that we have to check for this case twice -- once in the graph package and once in the dataplane package.

👍 

> What do you think about adding a Hostnames field to graph.Route? If an empty hostname is provided, the field will contain the wildcard hostname or the listener hostname; otherwise, it will be set to r.Source.Spec.Hostnames. Then in this function, we would only need one for loop that ranges over r.Hostnames. Would that work?

I don't think it will work if a route binds to multiple listeners.

Perhaps we can capture binded hostnames in graph.ParentRefAttachmentStatus ? 

ParentRefAttachmentStatus belongs to ParentRef which matches a parentRef in an HTTPRoute?

Reply:
This approach may also cause redundancy between the listener's `AcceptedHostnames` field and the route's `Hostnames` field.

Reply:
> This approach may also cause redundancy between the listener's AcceptedHostnames field and the route's Hostnames field.

what if we remove AcceptedHostnames?

Reply:
> I don't think it will work if a route binds to multiple listeners.

@pleshakov why not? We would still be cross-checking the listener's accepted hostnames. If a route has an empty hostname and attaches to two listeners -- one with hostname "example.com" and the other with "example.org" we will end up with:
```
route.Hostnames = ["example.com", "example.org"]
listener1.AcceptedHostnames = {"example.com": {} }
listener2.AcceptedHostnames = {"example.org": {} }
```

There will be some redundancy between the route's hostnames and the accepted hostnames, but they won't necessarily be the same.

Reply:
@kate-osborn 
thanks for the example. yep, it will work

 However, can we end up with this situation?
```
route.Hostnames = ["example.com", "example.com"]
listener1.AcceptedHostnames = {"example.com": {} } // http
listener2.AcceptedHostnames = {"example.com": {} } // https
```

Reply:
@pleshakov ah, yes that could happen. The route hostnames could be a set to prevent duplicates.

Reply:
Right now I'm exploring the option of removing the listener accepted hostnames, and adding an accepted hostnames map to the ParentRefAttachmentStatus, with one map entry per listener that's attached to the route.

I'll push the changes soon and we can re-review.

Reply:
Updates have been made to have the AcceptedHostnames map live on the Route's ParentRefAttachmentStatus rather than on the Listener. Each entry in the map is for an attached listener, with the value being the list of intersecting hostnames.

Reply:
LGTM!